bindings/rust/s2n-tls/src/connection.rs

```diff
@@ -64,6 +64,21 @@ impl fmt::Debug for Connection {
 /// s2n_connection objects can be sent across threads
 unsafe impl Send for Connection {}
 
+/// # Sync
+///
+/// Although NonNull isn't Sync and allows access to mutable pointers even from
+/// immutable references, the Connection interface enforces that all mutating
+/// methods correctly require &mut self.
+///
+/// Developers and reviewers MUST ensure that new methods correctly use
+/// either &self or &mut self depending on their behavior. No mechanism enforces this.
+///
+/// Note: Although non-mutating methods like getters should be thread-safe by definition,
+/// technically the only thread safety guarantee provided by the underlying C library
+/// is that s2n_send and s2n_recv can be called concurrently.
+///
```

Comment:
I feel like this PR is a good time to take a whack at #4413

Reply:
Added, but you may regret that ask because the PR is much larger now :)

Please make sure all my reasoning makes sense. I'm trying to walk a fine line between explaining what Send and Sync mean (NOT what our documentation should do-- there's a lot of Rust documentation specifically on that topic) and not being helpful at all.

Reply:
Actually, too much larger. I don't want discussion of the documentation changes to block the Sync fix. I've moved the documentation changes to another PR: https://github.com/aws/s2n-tls/pull/4471